### PR TITLE
[WIP] Add "wait" command for blocking until a job is finished

### LINKED ISSUE #427: Allow waiting for a woken-up job to finish
I'm using the [Local Snapshots + Offline Backup to an External Disk example](https://zrepl.github.io/quickstart/backup_to_external_disk.html) to backup a local pool to an external (i.e. **detachable**) disk.

I've created a script to decrypt the LUKS container on it, import the pool, wake up the push job ... well here my setup gets ugly: I've found no way to wait for the job I've just woken up to finish so I can export the pool and close the LUKS container cleanly.

Similar to [others]( https://github.com/zrepl/zrepl/issues/245#issuecomment-556016281) I also do brittle polling of `zrepl status --raw | jq 'magic'` in a loop until everything is done or something fails.

My feature suggestion:
Add a `-w` or `--wait` flag to `zrepl signal wakeup` that blocks until the job is done or something failed (exiting with non-zero code preferably :wink:). Alternatively there could be a `zrepl wait ...` subcommand.
FWIW: this would mimic behavior also present in OpenZFS. They also added `-w` flags to several long-running tasks as well as a [`zpool wait` command](https://openzfs.github.io/openzfs-docs/man/8/zpool-wait.8.html) .

### PR DESCRIPTION
**Pull request by riyad**
This is a rough first implementation for  #427.

Design decisions:
- calling the subcommand "wait"
- allow waiting only for one job

Non-descisions (i.e. there was no thought behind it):
- code structure and variable names mimic client/status
- sleep time between checks (currently 500msec)

Caveats:
- I'm writing Go for the first time
- I have relatively little experience with zrepl
- I only have a simple setup
- no (integration) tests

Todos:
- cover more corner cases ...
  -  .. through feedback 
  -  ... with tests
- better naming and code structure (any suggestions guidelines?)
- should there be meaningful error codes?
- should there be status/debug output?

Fixes  #427

### CODE REVIEW

File: client/wait.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+type wui struct {
+	//lock      sync.Mutex //For report and error
+	report map[string]*job.Status
+
+	jobName string
+}
+
+var WaitCmd = &cli.Subcommand{
+	Use:   "wait JOB",
+	Short: "block and wait until the specified job is done or has failed",
```
Comment (problame) [inline client/wait.go:28]: need to explain what `done` means somehwere

File: client/wait.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	Run: func(ctx context.Context, subcommand *cli.Subcommand, args []string) error {
+		return runWaitCmd(subcommand.Config(), args)
+	},
+}
+
+// adapted from status.go
```
Comment (problame) [inline client/wait.go:34]: perhaps we should re-use the `client` from the status-v2 work https://github.com/zrepl/zrepl/pull/297

File: client/wait.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	"github.com/zrepl/zrepl/cli"
+	"github.com/zrepl/zrepl/config"
+	"github.com/zrepl/zrepl/daemon"
+	"github.com/zrepl/zrepl/daemon/job"
+	"github.com/zrepl/zrepl/daemon/pruner"
+	"github.com/zrepl/zrepl/daemon/snapper"
+	"github.com/zrepl/zrepl/replication/report"
+)
+
+type wui struct {
```
Comment (problame) [inline client/wait.go:19]: confusing type name

File: client/wait.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+		done := replicationDone && senderPruningDone && receiverPruningDone
+
+		if v.Type == job.TypePush {
+			snapperDone, err := evaluateSnapperReport(activeStatus.Snapshotting)
+			if err != nil {
+				return true, err
+			}
+
+			done = done && snapperDone
```
Comment (problame) [inline client/wait.go:120]: Handling the snapshotting is tricky: snapshotting is never exactly 'done'. Maybe the command should be more like `zrepl wait-job JOBNAME --replication [true,false] --snapshotting [true,false] --pruning [true,false]` ?

See first commit in status-v2 `add signal 'snapshot', rename existing signal 'wakeup' to 'replication'` [link](https://github.com/zrepl/zrepl/pull/297/commits/a1a0c9f30552dbe2d42fd19a88e4dd40f88d8b57)

File: client/wait.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+	if r.Error != "" {
+		return true, fmt.Errorf("pruning error: %v", r.Error)
+	}
+
+	prunerState, err := pruner.StateString(r.State)
+	if err != nil {
+		return true, fmt.Errorf("parsing pruning state %v: %w", r.State, err)
+	}
+
+	switch prunerState {
```
Comment (problame) [inline client/wait.go:176]: we should have a `prunerState.IsTerminal()` getter, use it both here and in zrepl status

**Review by problame — COMMENTED** [review]
Thanks for putting this together.

I think this work would benefit from queuing up behind [status-v2](https://github.com/zrepl/zrepl/pull/297/commits), in particular the first commit that splits up signal `wakeup` into signals `replication` and signal `snapshot`.

**Comment by riyad** [comment]
Now that there's a new RC with the status-v2 changes, is it safe to start with a reimplementation of this feature?

File: client/wait.go (hunk 2 above)
Comment (riyad) [reply client/wait.go:34]: I'll look into it :slightly_smiling_face:

File: client/wait.go (hunk 3 above)
Comment (riyad) [reply client/wait.go:19]: yeah :sweat_smile:  ... did I mention it was adapted from the status code :wink:

File: client/wait.go (hunk 4 above)
Comment (riyad) [reply client/wait.go:120]: hmmm ... I was wondering if there're situations where a job finished fast enough that the wait after a signal would be stuck waiting on a job thinking that it's not started yet.

File: client/wait_job.go [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+			return true, err
+		}
+
+		snapperDone, err := evaluateSnapperStatus(snapStatus.Snapshotting)
+		if err != nil {
+			return true, err
+		}
+
+		return pruningDone && snapperDone, nil
+	} else if j.Type == job.TypeSource /*|| j.Type == job.TypeSink*/ {
```
Comment (riyad) [inline client/wait_job.go:130]: I'm not sure what to do here. The status code doesn't even mention `job.TypeSink`.

File: client/wait.go (hunk 4 above)
Comment (riyad) [reply client/wait.go]: I'd suggest keeping it simple at first and see if there's actually a need for sub-job granularity.

**Comment by problame** [comment]
> Now that there's a new RC with the status-v2 changes, is it safe to start with a reimplementation of this feature?

Sorry for not getting back to this sooner. @InsanePrawn and myself started drafting a new architecture for 'signal', 'reset', **and 'wakeup'** yesterday. The gist of the idea is to generalize the `invocation` id that we keep in `active.go` and use the tuple `(job, invocation)` to wait for a job to finish. This architecture will also allow for race-free `zrepl signal --wait JOB [replication|snapshotting]`, which we think is a pretty neat addition. Waiting for job completion is still poll based, but won't have to rely on polling the `status`.
Thus, I'd recommend waiting until next week - by then we'll hopefully have a draft PR out so we can see whether the design works out in practice.

Again, sorry for not getting back to this sooner :/

**Comment by riyad** [comment]
Awesome! tThis was a step up from shell magic and your solution is a huge step up from that. I'll take race-free any time. :+1:
This task was still a nice mental jog for me. :smile:
